validate.sh

```diff
@@ -0,0 +1,183 @@
+
+declare -a validate_objects=()
+
+validate_objects+=${uobjects[@]}
+validate_objects+=${queues[@]}
+validate_objects+=${stacks[@]}
+
+if [ "$1" = "--help" ] || [ "$1" = "-h" ]; then
+    usage;
+    exit;
+fi
+
+while [ "$1" != "" ]; do
+    PARAM=`echo $1 | awk -F= '{print $1}'`
+    VALUE=`echo $1 | awk -F= '{print $2}'`
+    case $PARAM in
+        -h | --help)
+            usage;
+            exit;
+            ;;
+        -s | --step_threads)
+            STEP_THREADS=$VALUE
+            STEP_SELETCTED=1
+            ;;
+        -t | --max_threads)
+            MAX_PTHREADS=$VALUE
+            ;;
+        -f | --fibers)
+            FIBERS="-f $VALUE"
+            ;;
+        -w | --max_work)
+            WORKLOAD="-w $VALUE"
+            ;;
+        -n | --numa_nodes)
+            NUMA_NODES="-n $VALUE"
+            ;;
+        -r | --runs)
+            RUNS_PER_THREAD=$VALUE
+            ;;
+        -*)
+            echo "ERROR: unknown parameter \"$PARAM\""
+            usage
+            exit 1
+            ;;
+	*)
+	    FILE=$PARAM
+	    ;;
+    esac
+    shift
+done
+
+# Calculate step according to user preferences
+if [ $STEP_SELETCTED -eq 0 ]; then
+    if [ $MAX_PTHREADS -gt 8 ]; then 
+        STEP_THREADS=$(($MAX_PTHREADS / 8))
+    fi
+fi
+
+# Fill the array with the selected number of threads
+PTHREADS_ARRAY=()
+
+# Add set of iterations for 1 thread, if needed
+if [ $STEP_THREADS -ne 1 ]; then
+    PTHREADS_ARRAY+=(1)
+fi
+# Add intermediate sets
+for (( PTHREADS=$STEP_THREADS; PTHREADS<$MAX_PTHREADS; PTHREADS+=$STEP_THREADS ));do
+    PTHREADS_ARRAY+=($PTHREADS)
+done
+# Add set of iterations for max threads
+PTHREADS_ARRAY+=($MAX_PTHREADS)
+
+echo "Compiling the sources..."
+make clean debug > $BUILD_LOG
+
+# Run the selected number of threads
+for PTHREADS in "${PTHREADS_ARRAY[@]}"; do
+    printf "\n\e[36mValidating for %3d threads\n" $PTHREADS
+    echo -e "==========================\e[39m"
+    for bench in "${uobjects[@]}"; do
+        printf "Validating %-20s \t\t\t\t\t" $bench
+        runs=$(($RUNS_PER_THREAD * $PTHREADS))
+        $BIN_PATH/$bench -t $PTHREADS -r $runs $WORKLOAD $FIBERS $NUMA_NODES > $RES_FILE 2>&1
+        state=$(fgrep "Object state: " $RES_FILE)
+        state=${state/#"DEBUG: Object state: "}
+        if [ $state -eq $runs ]; then
+            echo -e $COLOR_PASS
+        else
+            echo -e $COLOR_FAIL
+            echo "Expected state: " $runs
+            echo "Invalid state: " $state
+            PASS_STATUS=0
+        fi
+    done
+
+    for bench in "${stacks[@]}"; do
+        printf "Validating %-20s \t\t\t\t\t" $bench
+        runs=$(($RUNS_PER_THREAD * $PTHREADS))
+        $BIN_PATH/$bench -t $PTHREADS -r $runs $WORKLOAD $FIBERS $NUMA_NODES > $RES_FILE 2>&1
+        state=$(fgrep "Object state: " $RES_FILE)
+        state=${state/#"DEBUG: Object state: "}
+        runs=$(($runs * 2))
```

Comment:
which value is correct for the `runs` variable? This one or the one in line #133?

Reply:
In queues and stacks, the benchmarks execute pairs of enqueues/dequeues and push/pops respectively.
Thus, for queues and stacks the actual number of benchmark's operations is double.

Reply:
From what I see in every iteration we initialize the `runs` variable:
`runs=$(($RUNS_PER_THREAD * $PTHREADS))`
Then in this loop we double the stored value, and finally in the last loop we use the latest and greatest value of `runs`.
why this methodology?

Reply:
as discussed the issue is now resolved.